rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/WorkerBookKeeper.java

```diff
@@ -44,6 +92,38 @@ public void handleHeartbeat(String workerHostname, HeartbeatStatus request)
     throw new UnsupportedOperationException("Worker node should not handle heartbeat");
   }
 
+  @Override
+  public List<String> getNodeHostNames(int clusterType)
+  {
+    throw new UnsupportedOperationException("Worker node should not return list of nodes");
+  }
+
+  @Override
+  public String getClusterNodeHostName(String remotePath, int clusterType)
+  {
+    try {
+      initializeClusterManager(clusterType);
+    }
+    catch (ClusterManagerInitilizationException ex) {
+      log.error("Not able to initialize ClusterManager for cluster type : " + ClusterType.findByValue(clusterType) +
+          " with Exception : " + ex);
+      return null;
+    }
+
+    try {
+      List<String> nodes = nodeListCache.get(clusterType);
+      int nodeIndex = clusterManager.getNodeIndex(nodes.size(), remotePath);
+      String hostName = nodes.get(nodeIndex);
+
+      return hostName;
+    }
+    catch (ExecutionException e) {
+      e.printStackTrace();
+    }
+
+    return null;
```

Comment:
We can potentially return null from here but the consumer getCacheStatus() blindly puts the return value in map which will throw a NPE

Reply:
Good point. It is handled in CahingFileSystem.getFileBlockLocations but not in BookKeeper.getCacheStatus. Introduced a new block location called Unknown. If the hostname is null, we just read data from remote object store

Reply:
This is still not handled ... Adding 'if (blockSplits.get(split) == null) ` in BookKeeper.getCacheStatus will not be enough because you are still doing this there:
`      String hostName = getOwnerNodeForPath(key);
      blockSplits.put(blockNumber, hostName);
`

And Map.put(T, null) will throw NullPointerException

Reply:
`testGetCacheStatusWhenHostNameIsNull` proves we can put null as value in the map.